capi-matrix/README.md

```diff
@@ -0,0 +1,50 @@
+# RFC 0015 - CAPI matrix per provider
+
+## Introduction
+
+We have a cluster-<provider> chart for every provider that we support. It has a set of templates that allow our customers to create a cluster app with all necessary resources required by a Cluster API implementation. 
+
+In the values of these charts we define the Kubernetes version of the cluster and the image that machine will use to run. These images contain the Kubernetes components with a specific version. 
+
+One of the goal in this RFC is decouple Kubernetes version from image names so customer does not need to know which image correspond to which version. At the same time we might want to provide new OS version when new CVE appears without necessarily change the Kubernetes version. Finally having a list of kubernetes version certified as ready to be used benefit to disallow users to create cluster in flawy versions.
+
+## User stories
+
+- As a customer, I want to not worry about the kubernetes version and rely on Giant Swarm default version provided.
+
+- As Giant Swarm staff, I don't want to expend time upgrading cluster manually.
+
+- As Giant Swarm staff, I would like to forbid a specific Kubernetes/OS version to be used in case of security constraints.
+
+- As customer, I want to have the option to have a cluster pinned to a version.
+
+## Solution
+
+- Lets have a cluster-<provider> chart release for every Kubernetes version is released. That way our automation can roll upgrades automatically when new version is published
+
+- Lets include a matrix between Kubernetes version and images names in the chart.
+
+```yaml
+kubernetesVersionImages:
```

Comment:
Is this useful to the customer? How will that know what to pick? Why would we encourage the use of old (unsupported) Kubernetes versions?

Reply:
No this is not useful for the end customer (maybe there is a corner case) but maybe for other implementators like I describe at bottom of the doc. I have doubted to embedded in the _helpers but setting it values allow us to change it in real live if a critical issue appears

Reply:
I still don't really get why there needs to be a list of versions, if every new OS/K8s version is anyway a new chart release/version. Wouldn't each chart release then only need a single image?

I thought we pin an OS/K8s combination per chart release. This way the upgrade is moving to the next chart version, which is something that is easy to automate also later based on upstream automation instead of having to rely on value fields.

Only exception I'd see and that I'd like to support is having different images for CP and worker nodes, cause sometimes we might just want to update the CP nodes (CVE in K8s CP) without having to roll all workers.

Reply:
This would make the lifecycle of kubernetes independent from other features within the cluster chart. Eg independent of bastions, private clusters etc. But some other parts in the chart are still bound to kubernetes too. So if we want to have a team responsible for kubernetes itself and other teams responsible for infrastructure integration in a specific infrastructure like aws or openstack, how can we split responsibility within the same chart? Does the kubernetes team then still need to do releases of each provider chart? And is that a desired design? Are we even able to separate this completely? We need to make sure we separate the configuration for k8s itself and the provider one. The k8s team still needs to run tests on all providers and they need to have changes in their own chart dependent on which provider. But without the need to update all charts that are maintained by the infrastructure teams. They need to be able to roll out their own changes to all providers.

Another idea (not sure if discussed already) would be to allow customers to select something like beta, stable, long-term moving targets. This way teams can release their changes and the release gets tested further along and promoted through the stages.

Reply:
I agree it should be hidden from the customer. Did we ever had a critical issue requiring us to change the OS so quickly we wouldn't be able to change a few lines in the chart and release it?

Reply:
No I don't think so. It can be hidden of course, I have a discussion with someone else and making it part of values seemed logical for other possible users (not GS). 

And Timo I expected that if there is a K8s team releasing new version there is an automated process that creates the release for each provider (anyway so far is something we do manually, so this is not getting worse). Having multiple release channels could be a solution, but I would not make it hard to select a specific version of k8s/OS since it will could impact us in the future having to make workaround to pin versions.